src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/ProjectSystem/RazorProjectInfoEndpointPublisher.cs

```diff
@@ -0,0 +1,213 @@
+                {
+                    break;
+                }
+
+                EnqueuePublish(args.Newer.AssumeNotNull());
+
+                break;
+
+            case ProjectChangeKind.ProjectAdded:
+
+                if (ProjectWorkspacePublishable(args.Newer))
+                {
+                    // Don't enqueue project addition as those are unlikely to come through in large batches.
+                    // Also ensures that we won't get project removal go through the queue without project addition.
+                    ImmediatePublish(args.Newer.AssumeNotNull(), _disposeTokenSource.Token);
+                }
+
+                break;
+
+            case ProjectChangeKind.ProjectRemoved:
+                // Enqueue removal so it will replace any other project changes in the work queue as they unnecessary now.
+                EnqueueRemoval(args.Older.AssumeNotNull());
+                break;
+
+            default:
+                Debug.Fail("A new ProjectChangeKind has been added that the RazorProjectInfoEndpointPublisher doesn't know how to deal with");
+                break;
+        }
+    }
+
+    // We want to wait until our project has been initialized (ProjectWorkspaceState != null)
+    // so that we don't publish half-finished projects, which can cause things like Semantic coloring to "flash"
+    // when they update repeatedly as they load.
+    private static bool ProjectWorkspacePublishable(IProjectSnapshot? project)
+    {
+        return project?.ProjectWorkspaceState != null;
+    }
+
+    private void EnqueuePublish(IProjectSnapshot projectSnapshot)
+    {
+        _workQueue.AddWork((Project: projectSnapshot, Removal: false));
+    }
+
+    private void EnqueueRemoval(IProjectSnapshot projectSnapshot)
+    {
+        _workQueue.AddWork((Project: projectSnapshot, Removal: true));
+    }
+
+    private ValueTask ProcessBatchAsync(ImmutableArray<(IProjectSnapshot Project, bool Removal)> workItems, CancellationToken cancellationToken)
+    {
+        foreach (var workItem in workItems.GetMostRecentUniqueItems(Comparer.Instance))
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return default;
+            }
+
+            if (workItem.Removal)
+            {
+                RemovePublishingData(workItem.Project, cancellationToken);
+            }
+            else
+            {
+                ImmediatePublish(workItem.Project, cancellationToken);
+            }
+        }
+
+        return default;
+    }
+
+    private void RemovePublishingData(IProjectSnapshot projectSnapshot, CancellationToken cancellationToken)
+    {
+        // This should never get called if we are inactive, so don't check _active flag
+        ImmediatePublish(projectSnapshot.Key, encodedProjectInfo: null, cancellationToken);
+    }
+
+    private void ImmediatePublish(IProjectSnapshot projectSnapshot, CancellationToken cancellationToken)
+    {
+        if (!_active)
```

Comment:
I still think this field can be removed entirely. It's set before hooking up the event, and before publishing the initial project set, so will never be null

Reply:
You are right, it was a leftover from the original design, not needed anymore. Removed.